src/backend/distributed/commands/table.c

```diff
@@ -510,6 +510,20 @@ PreprocessAlterTableStmt(Node *node, const char *alterTableCommand,
 	bool referencingIsLocalTable = !IsCitusTable(leftRelationId);
 	if (referencingIsLocalTable)
 	{
+		if (list_length(alterTableStatement->cmds) == 1)
+		{
+			AlterTableCmd *command = linitial(alterTableStatement->cmds);
+			AlterTableType alterTableType = command->subtype;
+			if (alterTableType == AT_DropConstraint)
```

Comment:
Probably we need to consider some other `DROP X CASCADE` commands (like `DROP COLUMN`) as well since they might also drop such foreign keys when cascading, though it might extend scope of this bug fix.

Maybe one day, or today, we should decide to invalidate foreign key graph for any drop constraint command according to `ConstraintDropped` variable.

Reply:
Yes, invalidating foreign key graph for all drop constraint commands might solve this one, and other cases you mentioned. But could it have some side effects we don't want?

Reply:
> Maybe one day, or today, we should decide to invalidate foreign key graph for any drop constraint command according to ConstraintDropped variable.

This sounds like a solid idea, though lets not do it on the last day of release as you also noted :) 

I think at this moment, lets clean the current patch  (which makes sense as well) and commit&backport it.

```C
if (DropsForeignKey())
  {
/* some comment on why we need */
InvalidateForeignKeyGraph();
}
```

On Citus 10.1, we can improve the situation by relying on ConstraintDropped

Reply:
Thanks @onurctirtir for opening the issue https://github.com/citusdata/citus/issues/4676